sentence_transformers/losses/AdaptiveLayerLoss.py

```diff
@@ -0,0 +1,108 @@
+        Use the all_layer_embeddings to get the embeddings of all layers.
+        """
+        original_output_hidden_states = self.transformer.auto_model.config.output_hidden_states
+        self.transformer.auto_model.config.output_hidden_states = True
+
+        output = self.original_forward(features)
+        # We ignore the first layer, as it is the input embeddings
+        # and the last layer, as we already computed the loss over it
+        self.num_layers = len(output["all_layer_embeddings"]) - 1
+        self.embeddings.append(output["all_layer_embeddings"][1:-1])
+        self.last_embeddings.append(output["token_embeddings"])
+        self.features.append(
+            {key: value for key, value in output.items() if key not in ["all_layer_embeddings", "token_embeddings"]}
+        )
+
+        # Restore original setting
+        self.transformer.auto_model.config.output_hidden_states = original_output_hidden_states
+
+        if original_output_hidden_states:
+            del output["all_layer_embeddings"]
+
+        return output
+
+    def call_use_cache(self, features):
+        return {**self.features[self.call_idx], "token_embeddings": self.embeddings[self.call_idx][self.layer_idx]}
+
+
+class AdaptiveLayerLoss(nn.Module):
+    def __init__(
+        self,
+        model: SentenceTransformer,
+        loss: nn.Module,
+    ):
+        super().__init__()
+        self.model = model
+        self.loss = loss
+        assert isinstance(self.model[0], Transformer)
+
+    def forward(self, sentence_features: Iterable[Dict[str, Tensor]], labels: Tensor) -> Tensor:
+        original_forward = self.model[0].forward
+        decorated_forward = TransformerDecorator(self.model[0], original_forward)
+        self.model[0].forward = decorated_forward
+
+        # Run the loss normally: i.e. the final layer
+        # but use the decorator to get the embeddings of all layers
+        loss = self.loss(sentence_features, labels)
+        """
+        final_embeddings = torch.concat(decorated_forward.last_embeddings, dim=1)
+        """
+
+        # This loop is over `num_layer - 1` layers because we already computed the loss over the final layer
+        for layer_idx in range(decorated_forward.num_layers - 1):
+            # Add regular loss for each layer by using the cached embeddings of that layer
+            decorated_forward.set_layer_idx(layer_idx)
+            layer_loss = self.loss(sentence_features, labels) / (decorated_forward.num_layers - 1)
+            loss = loss + layer_loss
+
+            """
+            # and KL-divergence loss between the current layer and the final layer
+            # Note: we use "batchmean" reduction as that aligns with the mathematical definition
+            embeddings = decorated_forward.get_layer_embeddings()
+            kl_div_loss = F.kl_div(
+                F.log_softmax(embeddings, dim=-1), F.softmax(final_embeddings, dim=-1), reduction="batchmean"
```

Comment:
```suggestion
                F.log_softmax(embeddings, dim=-1), F.softmax(final_embeddings, dim=-1), reduction="batchmean"
```
I think the issue is that I'm running KL-div on the embeddings directly as they occur in the model, rather than after pooling etc. has been applied. 
Normally when you do knowledge distillation, you do the KL-div on the embeddings after pooling, and this approach is essentially knowledge distillation with the last embedding layer being your teacher model.

Reply:
I think this was it! 
![image](https://github.com/UKPLab/sentence-transformers/assets/37621491/2f2139fa-a84d-41c6-a085-c6a51793d367)
KL-div but after pooling is slightly better than no KL-div, just like in the paper! 🎉

Reply:
@tomaarsen Thanks for your excellent implementation! We used the pooling embeddings for KL, sorry for not pointing out this earlier.

However, I found that 2DMSE didn’t outperform MRL at the last layer. 
In my experiments, it consistently outperforms MRL at the last layer. I will try  the code tomorrow to see what caused this difference.

Reply:
> We used the pooling embeddings for KL, sorry for not pointing out this earlier.

No worries! I'm glad I figured it out quickly :)

> However, I found that 2DMSE didn’t outperform MRL at the last layer.
> In my experiments, it consistently outperforms MRL at the last layer.

Could you clarify this? This seems contradictory, unless I am understanding it wrong.

Reply:
In the first figure, I found MRL(the purple line) performs better than 2DMSE at the last layer. In my experiment, 2DMSE performed better than MRL at the last layer.

Reply:
Ahh, you're right. I think that might be because the purple line is the only model that uses the original MRL, the other models only use the "Adaptive Layer" part, not the MRL part. I just pushed some code to combine the two to create roughly 2DMSE (the only difference I believe is that my MatryoshkaLoss implementation uses all dimensions at once instead of sampling 1): 211eec32df0b06c49ec4bf850e4dcd78b0ff25c1

But I haven't tested with that one yet. I will do so soon & report back. Then hopefully the `Matryoshka2dLoss` will outperform `MatryoshkaLoss` at the last layer just like in your experiments.

Reply:
> Then hopefully the `Matryoshka2dLoss` will outperform `MatryoshkaLoss` at the last layer just like in your experiments.

No luck so far :/ 

`Matryoshka2dLoss` + `MNRL` at the final layer: 82.59
`Matryoshka2dLoss` + `MNRL` but with all layers at once (instead of sampling 1) at the final layer: 82.60
`AdaptiveLayerLoss` + `MNRL` at the final layer: 82.92
`MatryoshkaLoss` + `MNRL` at the final layer: 84.37
Just `MNRL` at the final layer: 83.06

So, MRL > 2DMSE at the final layer for me. `AdaptiveLayerLoss` + `MNRL` is about equivalent to just MNRL for me, which is nice. But `AdaptiveLayerLoss` + `MatryoshkaLoss` + `MNRL` (i.e. `Matryoshka2dLoss` + `MNRL`) is not equivalent to just `MatryoshkaLoss` + `MNRL` sadly.

I'll try more tests, this time with AnglELoss instead of MNRL. Perhaps that makes the difference

Reply:
![image](https://github.com/UKPLab/sentence-transformers/assets/37621491/5cde8c44-7e83-4381-b5a3-6770fddcdcf1)

This is with a version where I sample 1 dimension and 1 layer at a time, just like the real 2DMSE implementation (yellow). Already seems a bit better.

Reply:
Hi, I follow AnglE's implementation to add `temperature` to KL. Additionally, I set a random seed `42` for all models and used `CLS` pooling following the paper.

Here are the results:

- MatryoshkaLoss + CoSENT: 84.62
- AdaptiveLayerLoss (n_layers_per_step=1, temperature=1.0) + CoSENT: **84.71** 
- AdaptiveLayerLoss (n_layers_per_step=-1, temperature=1.0) + CoSENT: 84.58
- Matryoshka2dLoss (n_layers_per_step=1, temperature=0.3) + CoSENT: 84.53
- Matryoshka2dLoss (n_layers_per_step=-1, temperature=50.0) + CoSENT: 84.36

AdaptiveLayerLoss outperforms others, but Matryoshka2dLoss is still not as good as expected. I am still checking the differences between the two implementations.

 
<details>
<summary>train_sts.py</summary>

```python
"""
This examples trains BERT (or any other transformer model like RoBERTa, DistilBERT etc.) for the STSbenchmark from scratch. It generates sentence embeddings
that can be compared using cosine-similarity to measure the similarity.

Usage:
python training_nli.py

OR
python training_nli.py pretrained_transformer_model_name
"""
from torch.utils.data import DataLoader
import math
from sentence_transformers import SentenceTransformer, LoggingHandler, losses, models, util
from sentence_transformers.evaluation import EmbeddingSimilarityEvaluator
from sentence_transformers.readers import InputExample
import logging
from datetime import datetime
import sys
import os
import gzip
import csv
import torch
import numpy as np
import random

random_seed = 42
if random_seed is not None:
    os.environ['PYTHONHASHSEED'] = str(random_seed)
    random.seed(random_seed)
    np.random.seed(random_seed)
    torch.manual_seed(random_seed)

#### Just some code to print debug information to stdout
logging.basicConfig(
    format="%(asctime)s - %(message)s", datefmt="%Y-%m-%d %H:%M:%S", level=logging.INFO, handlers=[LoggingHandler()]
)
#### /print debug information to stdout


# Check if dataset exists. If not, download and extract  it
sts_dataset_path = "datasets/stsbenchmark.tsv.gz"

if not os.path.exists(sts_dataset_path):
    util.http_get("https://sbert.net/datasets/stsbenchmark.tsv.gz", sts_dataset_path)


# You can specify any huggingface/transformers pre-trained model here, for example, bert-base-uncased, roberta-base, xlm-roberta-base
model_name = sys.argv[1] if len(sys.argv) > 1 else "distilbert-base-uncased"

# Read the dataset
train_batch_size = 16
num_epochs = 4
model_save_path = (
    "output/training_stsbenchmark_" + model_name.replace("/", "-") + "-" + datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
)

# Use Huggingface/transformers model (like BERT, RoBERTa, XLNet, XLM-R) for mapping tokens to embeddings
word_embedding_model = models.Transformer(model_name)

# Apply mean pooling to get one fixed sized sentence vector
pooling_model = models.Pooling(
    word_embedding_model.get_word_embedding_dimension(),
    pooling_mode_mean_tokens=False,
    pooling_mode_cls_token=True,
    pooling_mode_max_tokens=False,
)

model = SentenceTransformer(modules=[word_embedding_model, pooling_model])

# Convert the dataset to a DataLoader ready for training
logging.info("Read STSbenchmark train dataset")

train_samples = []
dev_samples = []
test_samples = []
with gzip.open(sts_dataset_path, "rt", encoding="utf8") as fIn:
    reader = csv.DictReader(fIn, delimiter="\t", quoting=csv.QUOTE_NONE)
    for row in reader:
        score = float(row["score"]) / 5.0  # Normalize score to range 0 ... 1
        inp_example = InputExample(texts=[row["sentence1"], row["sentence2"]], label=score)

        if row["split"] == "dev":
            dev_samples.append(inp_example)
        elif row["split"] == "test":
            test_samples.append(inp_example)
        else:
            train_samples.append(inp_example)


train_dataloader = DataLoader(train_samples, shuffle=True, batch_size=train_batch_size)
train_loss = losses.CoSENTLoss(model)
# MatryoshkaLoss + CoSENT: 84.62
# AdaptiveLayerLoss (n_layers_per_step=1, temperature=1.0) + CoSENT: **84.71** 
# AdaptiveLayerLoss (n_layers_per_step=-1, temperature=1.0) + CoSENT: 84.58
# Matryoshka2dLoss (n_layers_per_step=1, temperature=0.3) + CoSENT: 84.53
# Matryoshka2dLoss (n_layers_per_step=-1, temperature=50.0) + CoSENT: 84.36
hidden_sizes = [768, 512, 256, 128, 64, 32, 16, 8]
# train_loss = losses.MatryoshkaLoss(model, train_loss, hidden_sizes)
# train_loss = losses.AdaptiveLayerLoss(model, train_loss, n_layers_per_step=1)
train_loss = losses.Matryoshka2dLoss(model, train_loss, hidden_sizes, n_layers_per_step=-1, n_dims_per_step=1, temperature=0.3)
print('train loss=', train_loss)

logging.info("Read STSbenchmark dev dataset")
evaluator = EmbeddingSimilarityEvaluator.from_input_examples(dev_samples, name="sts-dev") # dev_samples


# Configure the training. We skip evaluation in this example
warmup_steps = math.ceil(len(train_dataloader) * num_epochs * 0.1)  # 10% of train data for warm-up
logging.info("Warmup-steps: {}".format(warmup_steps))


# Train the model
model.fit(
    train_objectives=[(train_dataloader, train_loss)],
    evaluator=evaluator,
    epochs=num_epochs,
    evaluation_steps=1000,
    warmup_steps=warmup_steps,
    output_path=model_save_path,
)


##############################################################################
#
# Load the stored model and evaluate its performance on STS benchmark dataset
#
##############################################################################

model = SentenceTransformer(model_save_path)
test_evaluator = EmbeddingSimilarityEvaluator.from_input_examples(test_samples, name="sts-test")
test_evaluator(model, output_path=model_save_path)
```
</details>

<details>
<summary>Matryoshka2dLoss.py</summary>

```python
from typing import List, Optional, Union
from torch.nn import Module
from sentence_transformers.SentenceTransformer import SentenceTransformer
from sentence_transformers.losses import AdaptiveLayerLoss, MatryoshkaLoss


class Matryoshka2dLoss(AdaptiveLayerLoss):
    def __init__(
        self,
        model: SentenceTransformer,
        loss: Module,
        matryoshka_dims: List[int],
        matryoshka_weights: Optional[List[Union[float, int]]] = None,
        n_layers_per_step: int = 1,
        n_dims_per_step: int = 1,
        temperature: float = 1.0,
    ) -> None:
        # Note, this uses n_layers_per_step=1 & n_dims_per_step=1 as default, following the 2DMSE implementation
        matryoshka_loss = MatryoshkaLoss(
            model, loss, matryoshka_dims, matryoshka_weights=matryoshka_weights, n_dims_per_step=n_dims_per_step
        )
        super().__init__(model, matryoshka_loss, n_layers_per_step, temperature=temperature)
```
</details>

<details>
<summary>AdaptiveLayerLoss.py</summary>

```python
import random
from typing import Dict, Iterable, List, Tuple
import warnings
from torch import Tensor, nn
from torch.nn import functional as F
import torch
from sentence_transformers import SentenceTransformer
from sentence_transformers.losses.CachedMultipleNegativesRankingLoss import CachedMultipleNegativesRankingLoss
from sentence_transformers.models import Transformer


class TransformerDecorator:
    """
    Decorator that caches the embeddings of all layers of the transformer.
    When `layer_idx` is set, it returns the cached embeddings of that layer instead.

    This is meant to override the forward function of the Transformer.
    """

    def __init__(self, transformer: Transformer, original_forward):
        self.transformer = transformer
        self.original_forward = original_forward
        self.embeddings: List[Tuple[Tensor]] = []
        self.last_embeddings: List[Tensor] = []
        self.features: List[Dict[str, Tensor]] = []
        self.layer_idx = None
        self.call_idx = 0

    def set_layer_idx(self, layer_idx):
        self.layer_idx = layer_idx
        self.call_idx = 0

    def get_layer_embeddings(self):
        return torch.concat([embedding[self.layer_idx] for embedding in self.embeddings], dim=1)

    def __call__(self, features):
        if self.layer_idx is None:
            output = self.call_grow_cache(features)
        else:
            output = self.call_use_cache(features)
            self.call_idx += 1
        return output

    def call_grow_cache(self, features):
        """
        Temporarily sets the output_hidden_states to True, runs the model, and then restores the original setting.
        Use the all_layer_embeddings to get the embeddings of all layers.
        """
        original_output_hidden_states = self.transformer.auto_model.config.output_hidden_states
        self.transformer.auto_model.config.output_hidden_states = True

        output = self.original_forward(features)
        # We ignore the first layer, as it is the input embeddings
        # and the last layer, as we already computed the loss over it
        self.num_layers = len(output["all_layer_embeddings"]) - 1
        self.embeddings.append(output["all_layer_embeddings"][1:-1])
        self.last_embeddings.append(output["token_embeddings"])
        self.features.append(
            {key: value for key, value in output.items() if key not in ["all_layer_embeddings", "token_embeddings"]}
        )

        # Restore original setting
        self.transformer.auto_model.config.output_hidden_states = original_output_hidden_states

        if original_output_hidden_states:
            del output["all_layer_embeddings"]

        return output

    def call_use_cache(self, features):
        return {**self.features[self.call_idx], "token_embeddings": self.embeddings[self.call_idx][self.layer_idx]}


class ForwardDecorator:
    """
    Decorator that caches the embeddings after all modules (e.g. pooling) of the model.
    Required to get the embeddings after all modules for the KL-divergence loss.

    This is meant to override the forward function of the SentenceTransformer.
    """

    def __init__(self, fn):
        self.fn = fn
        self.embeddings = []

    def __call__(self, features):
        output = self.fn(features)
        self.embeddings.append(output["sentence_embedding"])
        return output

    def get_embeddings(self):
        embeddings = torch.concat(self.embeddings, dim=0)
        self.embeddings = []
        return embeddings


class AdaptiveLayerLoss(nn.Module):
    def __init__(
        self,
        model: SentenceTransformer,
        loss: nn.Module,
        n_layers_per_step: int = -1,
        temperature: float = 1.0,
    ) -> None:
        """
        The AdaptiveLayerLoss can be seen as a loss *modifier* that allows you to use other loss functions at non-final
        layers of the Sentence Transformer model. This is useful for when you want to train a model where users have
        the option to lower the number of layers used to improve their inference speed and memory usage.

        :param model: SentenceTransformer model
        :param loss: The loss function to be used, e.g. :class:`MultipleNegativesRankingLoss`, :class:`CoSENTLoss`, etc.
        :param n_layers_per_step: The number of layers to use per step. If -1, then all layers are used. If > 0, then
            a random sample of n_layers_per_step layers are used per step. The 2DMSE paper uses `n_layers_per_step=1`.
            The default value is -1.

        References:
            - The concept was inspired by the 2DMSE paper: https://arxiv.org/abs/2402.14776

        Requirements:
            1. The base loss cannot be :class:`CachedMultipleNegativesRankingLoss`.

        Input:
            +---------------------------------------+--------+
            | Texts                                 | Labels |
            +=======================================+========+
            | any                                   | any    |
            +---------------------------------------+--------+

        Example:
            ::

                from sentence_transformers import SentenceTransformer, losses, InputExample
                from torch.utils.data import DataLoader

                model = SentenceTransformer('microsoft/mpnet-base')
                train_examples = [
                    InputExample(texts=['Anchor 1', 'Positive 1']),
                    InputExample(texts=['Anchor 2', 'Positive 2']),
                ]
                train_dataloader = DataLoader(train_examples, shuffle=True, batch_size=32)
                train_loss = losses.MultipleNegativesRankingLoss(model=model)
                train_loss = losses.AdaptiveLayerLoss(model, train_loss)
                model.fit(
                    [(train_dataloader, train_loss)],
                    epochs=10,
                )
        """
        super().__init__()
        self.model = model
        self.loss = loss
        self.n_layers_per_step = n_layers_per_step
        self.temperature = temperature
        assert isinstance(self.model[0], Transformer)
        if isinstance(loss, CachedMultipleNegativesRankingLoss):
            warnings.warn("MatryoshkaLoss is not compatible with CachedMultipleNegativesRankingLoss.", stacklevel=2)

    def forward(self, sentence_features: Iterable[Dict[str, Tensor]], labels: Tensor) -> Tensor:
        # Decorate the forward function of the transformer to cache the embeddings of all layers
        original_transformer_forward = self.model[0].forward
        transformer_decorator = TransformerDecorator(self.model[0], original_transformer_forward)
        self.model[0].forward = transformer_decorator

        # Decorate the forward function of the model to get the embeddings after all modules (e.g. pooling)
        original_forward = self.model.forward
        forward_decorator = ForwardDecorator(original_forward)
        self.model.forward = forward_decorator

        # Run the loss normally: i.e. the final layer, but 1) use the transformers decorator to cache
        # the embeddings of all layers and 2) use the forward decorator to get the embeddings after all modules
        # for the KL-divergence loss
        loss = self.loss(sentence_features, labels)
        final_embeddings = forward_decorator.get_embeddings()
        final_embeddings = F.softmax(final_embeddings / self.temperature, dim=-1)

        num_layers = transformer_decorator.num_layers
        layer_indices = range(num_layers - 1)
        if self.n_layers_per_step > 0 and self.n_layers_per_step < num_layers - 1:
            layer_indices = random.sample(layer_indices, self.n_layers_per_step)

        # This loop is over `num_layer - 1` layers because we already computed the loss over the final layer
        for layer_idx in layer_indices:
            # Add regular loss for each layer by using the cached embeddings of that layer
            transformer_decorator.set_layer_idx(layer_idx)
            layer_loss = self.loss(sentence_features, labels)
            loss = loss + layer_loss / len(layer_indices)

            # and KL-divergence loss between the current layer and the final layer
            # Note: we use "batchmean" reduction as that aligns with the mathematical definition
            embeddings = forward_decorator.get_embeddings()
            kl_div_loss = F.kl_div(F.log_softmax(embeddings / self.temperature, dim=-1),
                                   final_embeddings, reduction="batchmean")
            loss = loss + kl_div_loss * self.temperature / len(layer_indices)

        self.model[0].forward = original_transformer_forward
        self.model.forward = original_forward

        return loss
```
</detail>

Reply:
Thanks a bunch for these experiments! What temperature did you use in the paper? 
Also, 
```python
loss = loss + kl_div_loss * self.temperature / len(layer_indices)
```
Should this be 
```python
loss = loss + kl_div_loss * self.temperature**2 / len(layer_indices)
```
perhaps? I tried the latter, but it performed a bit worse (in my few experiments).

Also, I realised another difference. My `MatryoshkaLoss` with `self.n_dims_per_step=1` just samples 1 dimension, rather than 1 dimension *plus* the largest dimension.

Reply:
I've updated the implementation of AdaptiveLayerLoss.py, here are the new results:
Now, 2dmse outperforms MRL using the same hyperparameters.

**random seed 42**
- MatryoshkaLoss + CoSENT: 84.62
- AdaptiveLayerLoss (n_layers_per_step=1, temperature=1.0) + CoSENT: 85.17
- AdaptiveLayerLoss (n_layers_per_step=-1, temperature=1.0) + CoSENT: **85.36**
- Matryoshka2dLoss (n_layers_per_step=1, temperature=0.3) + CoSENT: 84.73
- Matryoshka2dLoss (n_layers_per_step=-1, temperature=0.3) + CoSENT: 84.11

<details>
<summary>AdaptiveLayerLoss.py</summary>

Main Diffs:

1)
```python
loss = loss + layer_loss / len(layer_indices)
```
-> 
```python
loss = loss + layer_loss / (1 + layer_idx) / len(layer_indices)
```

2)
```python
loss = loss + kl_div_loss * self.temperature / len(layer_indices)
```
->
```python
loss = loss + kl_div_loss * self.temperature
```

```python
import math
import random
from typing import Dict, Iterable, List, Tuple
import warnings
from torch import Tensor, nn
from torch.nn import functional as F
import torch
from sentence_transformers import SentenceTransformer
from sentence_transformers.losses.CachedMultipleNegativesRankingLoss import CachedMultipleNegativesRankingLoss
from sentence_transformers.models import Transformer


class TransformerDecorator:
    """
    Decorator that caches the embeddings of all layers of the transformer.
    When `layer_idx` is set, it returns the cached embeddings of that layer instead.

    This is meant to override the forward function of the Transformer.
    """

    def __init__(self, transformer: Transformer, original_forward):
        self.transformer = transformer
        self.original_forward = original_forward
        self.embeddings: List[Tuple[Tensor]] = []
        self.last_embeddings: List[Tensor] = []
        self.features: List[Dict[str, Tensor]] = []
        self.layer_idx = None
        self.call_idx = 0

    def set_layer_idx(self, layer_idx):
        self.layer_idx = layer_idx
        self.call_idx = 0

    def get_layer_embeddings(self):
        return torch.concat([embedding[self.layer_idx] for embedding in self.embeddings], dim=1)

    def __call__(self, features):
        if self.layer_idx is None:
            output = self.call_grow_cache(features)
        else:
            output = self.call_use_cache(features)
            self.call_idx += 1
        return output

    def call_grow_cache(self, features):
        """
        Temporarily sets the output_hidden_states to True, runs the model, and then restores the original setting.
        Use the all_layer_embeddings to get the embeddings of all layers.
        """
        original_output_hidden_states = self.transformer.auto_model.config.output_hidden_states
        self.transformer.auto_model.config.output_hidden_states = True

        output = self.original_forward(features)
        # We ignore the first layer, as it is the input embeddings
        # and the last layer, as we already computed the loss over it
        self.num_layers = len(output["all_layer_embeddings"]) - 1
        self.embeddings.append(output["all_layer_embeddings"][1:-1])
        self.last_embeddings.append(output["token_embeddings"])
        self.features.append(
            {key: value for key, value in output.items() if key not in ["all_layer_embeddings", "token_embeddings"]}
        )

        # Restore original setting
        self.transformer.auto_model.config.output_hidden_states = original_output_hidden_states

        if original_output_hidden_states:
            del output["all_layer_embeddings"]

        return output

    def call_use_cache(self, features):
        return {**self.features[self.call_idx], "token_embeddings": self.embeddings[self.call_idx][self.layer_idx]}


class ForwardDecorator:
    """
    Decorator that caches the embeddings after all modules (e.g. pooling) of the model.
    Required to get the embeddings after all modules for the KL-divergence loss.

    This is meant to override the forward function of the SentenceTransformer.
    """

    def __init__(self, fn):
        self.fn = fn
        self.embeddings = []

    def __call__(self, features):
        output = self.fn(features)
        self.embeddings.append(output["sentence_embedding"])
        return output

    def get_embeddings(self):
        embeddings = torch.concat(self.embeddings, dim=0)
        self.embeddings = []
        return embeddings


class AdaptiveLayerLoss(nn.Module):
    def __init__(
        self,
        model: SentenceTransformer,
        loss: nn.Module,
        n_layers_per_step: int = -1,
        temperature: float = 1.0,
    ) -> None:
        """
        The AdaptiveLayerLoss can be seen as a loss *modifier* that allows you to use other loss functions at non-final
        layers of the Sentence Transformer model. This is useful for when you want to train a model where users have
        the option to lower the number of layers used to improve their inference speed and memory usage.

        :param model: SentenceTransformer model
        :param loss: The loss function to be used, e.g. :class:`MultipleNegativesRankingLoss`, :class:`CoSENTLoss`, etc.
        :param n_layers_per_step: The number of layers to use per step. If -1, then all layers are used. If > 0, then
            a random sample of n_layers_per_step layers are used per step. The 2DMSE paper uses `n_layers_per_step=1`.
            The default value is -1.

        References:
            - The concept was inspired by the 2DMSE paper: https://arxiv.org/abs/2402.14776

        Requirements:
            1. The base loss cannot be :class:`CachedMultipleNegativesRankingLoss`.

        Input:
            +---------------------------------------+--------+
            | Texts                                 | Labels |
            +=======================================+========+
            | any                                   | any    |
            +---------------------------------------+--------+

        Example:
            ::

                from sentence_transformers import SentenceTransformer, losses, InputExample
                from torch.utils.data import DataLoader

                model = SentenceTransformer('microsoft/mpnet-base')
                train_examples = [
                    InputExample(texts=['Anchor 1', 'Positive 1']),
                    InputExample(texts=['Anchor 2', 'Positive 2']),
                ]
                train_dataloader = DataLoader(train_examples, shuffle=True, batch_size=32)
                train_loss = losses.MultipleNegativesRankingLoss(model=model)
                train_loss = losses.AdaptiveLayerLoss(model, train_loss)
                model.fit(
                    [(train_dataloader, train_loss)],
                    epochs=10,
                )
        """
        super().__init__()
        self.model = model
        self.loss = loss
        self.n_layers_per_step = n_layers_per_step
        self.temperature = temperature
        print('n_layers_per_step>>>', n_layers_per_step)
        print('temperature2>>>', temperature)
        assert isinstance(self.model[0], Transformer)
        if isinstance(loss, CachedMultipleNegativesRankingLoss):
            warnings.warn("MatryoshkaLoss is not compatible with CachedMultipleNegativesRankingLoss.", stacklevel=2)

    def forward(self, sentence_features: Iterable[Dict[str, Tensor]], labels: Tensor) -> Tensor:
        # Decorate the forward function of the transformer to cache the embeddings of all layers
        original_transformer_forward = self.model[0].forward
        transformer_decorator = TransformerDecorator(self.model[0], original_transformer_forward)
        self.model[0].forward = transformer_decorator

        # Decorate the forward function of the model to get the embeddings after all modules (e.g. pooling)
        original_forward = self.model.forward
        forward_decorator = ForwardDecorator(original_forward)
        self.model.forward = forward_decorator

        # Run the loss normally: i.e. the final layer, but 1) use the transformers decorator to cache
        # the embeddings of all layers and 2) use the forward decorator to get the embeddings after all modules
        # for the KL-divergence loss
        loss = self.loss(sentence_features, labels)
        final_embeddings = forward_decorator.get_embeddings()
        final_embeddings = F.softmax(final_embeddings / self.temperature, dim=-1)

        num_layers = transformer_decorator.num_layers
        layer_indices = range(num_layers - 1)
        if self.n_layers_per_step > 0 and self.n_layers_per_step < num_layers - 1:
            layer_indices = random.sample(layer_indices, self.n_layers_per_step)

        # This loop is over `num_layer - 1` layers because we already computed the loss over the final layer
        for layer_idx in layer_indices:
            # Add regular loss for each layer by using the cached embeddings of that layer
            transformer_decorator.set_layer_idx(layer_idx)
            layer_loss = self.loss(sentence_features, labels)
            loss = loss + layer_loss / (1 + layer_idx) / len(layer_indices)

            # and KL-divergence loss between the current layer and the final layer
            # Note: we use "batchmean" reduction as that aligns with the mathematical definition
            embeddings = forward_decorator.get_embeddings()
            kl_div_loss = F.kl_div(F.log_softmax(embeddings / self.temperature, dim=-1),
                                   final_embeddings, reduction="batchmean")
            loss = loss + kl_div_loss * self.temperature

        self.model[0].forward = original_transformer_forward
        self.model.forward = original_forward

        return loss

```
</details>